cli/internal/runsummary/spaces.go

```diff
@@ -1,13 +1,258 @@
 package runsummary
 
 import (
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/mitchellh/cli"
 	"github.com/vercel/turbo/cli/internal/ci"
+	"github.com/vercel/turbo/cli/internal/client"
 )
 
-// spacesRunResponse deserialized the response from POST Run endpoint
-type spacesRunResponse struct {
-	ID  string
-	URL string
+const runsEndpoint = "/v0/spaces/%s/runs"
+const runsPatchEndpoint = "/v0/spaces/%s/runs/%s"
+const tasksEndpoint = "/v0/spaces/%s/runs/%s/tasks"
+
+// spaceRequest contains all the information for a single request to Spaces
+type spaceRequest struct {
+	method  string
+	url     string
+	body    interface{}
+	makeURL func(self *spaceRequest, r *spaceRun) error // Should set url on self
+	onDone  func(self *spaceRequest, response []byte)   // Handler for when request completes
+}
+
+func (req *spaceRequest) error(msg string) error {
+	return fmt.Errorf("[%s] %s: %s", req.method, req.url, msg)
+}
+
+type spacesClient struct {
+	requests chan *spaceRequest
+	errors   []error
+	api      *client.APIClient
+	ui       cli.Ui
+	run      *spaceRun
+	wg       sync.WaitGroup
+	spaceID  string
+	enabled  bool
+}
+
+type spaceRun struct {
+	ID      string
+	URL     string
+	created chan struct{} // a signal that the run has completed
+}
+
+func newSpacesClient(spaceID string, api *client.APIClient, ui cli.Ui) *spacesClient {
+	return &spacesClient{
+		api:      api,
+		ui:       ui,
+		spaceID:  spaceID,
+		enabled:  spaceID != "",
+		requests: make(chan *spaceRequest), // TODO: give this a size based on tasks
+		// Set a default, empty one here, so we'll have something downstream and not a segfault
+		run: &spaceRun{
```

Comment:
This, combined with the `Unmarshal` call below, makes me a bit nervous. It appears to be mixing resources and data (channel and `ID`, `URL`), as well as potentially using uninitialized zero-values. It might be better to split the struct into resource management that needs lifecycle control (creating, closing the channel) and plain data (`ID`, `URL`). You shouldn't need extra concern about segfaulting because it should be the same logic as making sure you have valid values before using them. I _think_ in this case it's the same as not using the data until `created` is closed?

Reply:
separated the two